gateway-ha/src/main/java/io/trino/gateway/ha/router/QueryCountBasedRouter.java

```diff
@@ -0,0 +1,151 @@
+ * limitations under the License.
+ */
+
+package io.trino.gateway.ha.router;
+
+import io.trino.gateway.ha.clustermonitor.ClusterStats;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+import java.util.stream.Collectors;
+
+public class QueryCountBasedRouter
+            extends HaRoutingManager
+{
+    private static final Logger log = LoggerFactory.getLogger(QueryCountBasedRouter.class);
+    private ArrayList<ClusterStats> clusterStats;
+    private static final ReentrantReadWriteLock lock = new ReentrantReadWriteLock(true);
+
+    public QueryCountBasedRouter(GatewayBackendManager gatewayBackendManager,
+                                 QueryHistoryManager queryHistoryManager)
+    {
+        super(gatewayBackendManager, queryHistoryManager);
+        clusterStats = new ArrayList<>();
+    }
+
+    // We sort and find the backend based on the individual user's count of the queued queries
+    // first, in case user doesn't have any queries queued we use the cluster wide stats
+    //
+    // First filter the list of clusters for a particular routing group
+    //
+    // If a user has queued queries, then find a cluster with the least number of QUEUED queries
+    // for that user.
+    //
+    // If a user's QUEUED query count is the same on every cluster then go with a cluster with
+    // cluster wide stats.
+    //
+    // Find a cluster with the least number of QUEUED queries, if there are the same number of
+    // queries queued, then compare the number of running queries.
+    //
+    // After a query is routed, we need to update the stats for that cluster until we received the
+    // updated stats for all the clusters.
+    // if a user has queries queued then we assume that the routed query will be also queued or
+    // else we assume it would be scheduled immediately and we increment the stats for the running
+    // queries
+
+    private Optional<ClusterStats> getClusterToRoute(String user, String routingGroup)
+    {
+        List<ClusterStats> filteredList = null;
+        Lock readLock = lock.readLock();
+        try {
+            readLock.lock();
+            log.debug("sorting cluster stats for {} {}", user, routingGroup);
+            filteredList = clusterStats.stream()
+                .filter(stats -> stats.healthy())
+                .filter(stats -> routingGroup.equals(stats.routingGroup()))
+                .collect(Collectors.toList());
+        }
+        finally {
+            readLock.unlock();
+        }
+
+        if (filteredList.isEmpty()) {
+            return Optional.empty();
+        }
+
+        Collections.sort(filteredList, new Comparator<ClusterStats>()
+        {
+            public int compare(ClusterStats lhs, ClusterStats rhs)
```

Comment:
`ClusterStats` should extend `Comparable`, then you can implement `compareTo` using this logic. That way it can be reused in other routers.

Reply:
That's true!

Reply:
looks good, can you remove the comment block of the previous code?